Copyright (c) Huawei Technologies Co., Ltd. 2021-2021. All rights reserved.


diff --git a/ojluni/src/main/native/UNIXProcess_md.c b/ojluni/src/main/native/UNIXProcess_md.c
index 550a5c5932..80370bf6b6 100644
--- a/ojluni/src/main/native/UNIXProcess_md.c
+++ b/ojluni/src/main/native/UNIXProcess_md.c
@@ -118,10 +118,12 @@
 // Android-changed: disable vfork under AddressSanitizer.
 //  #ifdef __linux__
   #if defined(__linux__) && !__has_feature(address_sanitizer)
-    #define START_CHILD_USE_VFORK 1
+    #define START_CHILD_USE_VFORK 0
   #else
     #define START_CHILD_USE_VFORK 0
   #endif
+#else
+#define START_CHILD_USE_VFORK 0
 #endif
 
 #if START_CHILD_USE_CLONE
